packages/react/src/use-chat.ts

```diff
@@ -183,45 +190,68 @@ Default is undefined, which disables throttling.
   const chatId = id ?? hookId;
   const chatKey = typeof api === 'string' ? [api, chatId] : chatId;
 
-  // Store array of the processed initial messages to avoid re-renders:
-  const stableInitialMessages = useStableValue(initialMessages ?? []);
-  const processedInitialMessages = useMemo(
-    () => stableInitialMessages,
-    [stableInitialMessages],
+  const chatStore = useRef(
+    store ??
+      new ChatStore({
+        chats: {
+          [chatId]: {
+            messages: [],
+          },
+        },
+        '~internal': {
+          currentDate: getCurrentDate,
+        },
+      }),
   );
 
-  // Store the chat state in SWR, using the chatId as the key to share states.
-  const { data: messages, mutate } = useSWR<UIMessage[]>(
-    [chatKey, 'messages'],
-    null,
-    { fallbackData: processedInitialMessages },
-  );
+  if (!chatStore.current.hasChat(chatId)) {
+    chatStore.current.addChat(chatId, []);
+  }
 
-  // Keep the latest messages in a ref.
-  const messagesRef = useRef<UIMessage[]>(messages || []);
-  useEffect(() => {
-    messagesRef.current = messages || [];
-  }, [messages]);
+  const { subscribe, getMessages, getStatus, getError, setStatus } =
+    useChatStore({
+      store: chatStore.current,
+    });
+
+  const messages = useSyncExternalStore(
+    callback =>
+      subscribe({
+        onStoreChange: callback,
+        eventType: 'chat-messages-changed',
+      }),
+    () => getMessages(chatId),
```

Comment:
Are the messages immutable? (even deeper in the tree, e.g. some tool invocations)

Reply:
they are not _truly_ immutable + should be treated as read-only (which we can add to documentation); only mutated via methods on chat store! from our end, to truly enforce immutability, we'd likely want to introduce something like `immer` - would love to discuss this further tomorrow!